sha3/src/cshake.rs

```diff
@@ -0,0 +1,200 @@
+                );
+                buffer.digest_blocks(customization, |blocks| state.update_blocks(blocks));
+                state.update_blocks(&[buffer.pad_with_zeros()]);
+                state.initial_state = state.state;
+                state
+            }
+        }
+
+        impl CustomizedInit for $name {
+            #[inline]
+            fn new_customized(customization: &[u8]) -> Self {
+                Self::new_with_function_name(&[], customization)
+            }
+        }
+
+        impl BufferKindUser for $name {
+            type BufferKind = Eager;
+        }
+
+        impl HashMarker for $name {}
+
+        impl BlockSizeUser for $name {
+            type BlockSize = $rate;
+        }
+
+        impl UpdateCore for $name {
+            #[inline]
+            fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+                for block in blocks {
+                    xor_block(&mut self.state, block);
+                    keccak::p1600(&mut self.state, ROUNDS);
+                }
+            }
+        }
+
+        impl ExtendableOutputCore for $name {
+            type ReaderCore = Sha3XofReaderCore<$rate>;
+
+            #[inline]
+            fn finalize_xof_core(&mut self, buffer: &mut Buffer<Self>) -> Self::ReaderCore {
+                let pos = buffer.get_pos();
+                let mut block = buffer.pad_with_zeros();
+                let pad = if self.initial_state == [0; PLEN] {
+                    SHAKE_PAD
+                } else {
+                    CSHAKE_PAD
+                };
+                block[pos] = pad;
+                let n = block.len();
+                block[n - 1] |= 0x80;
+
+                xor_block(&mut self.state, &block);
+                keccak::p1600(&mut self.state, ROUNDS);
+
+                Sha3XofReaderCore::new(&self.state)
+            }
+        }
+
+        impl Reset for $name {
+            #[inline]
+            fn reset(&mut self) {
+                self.state = self.initial_state;
+            }
+        }
+
+        impl AlgorithmName for $name {
+            fn write_alg_name(f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str($alg_name)
+            }
+        }
+
+        impl fmt::Debug for $name {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str(concat!(stringify!($name), " { ... }"))
+            }
+        }
+
+        impl Drop for $name {
+            fn drop(&mut self) {
+                #[cfg(feature = "zeroize")]
+                {
+                    use digest::zeroize::Zeroize;
+                    self.state.zeroize();
+                    self.initial_state.zeroize();
+                }
+            }
+        }
+
+        #[cfg(feature = "zeroize")]
+        impl digest::zeroize::ZeroizeOnDrop for $name {}
+
+        impl SerializableState for $name {
+            // TODO: implement properly with U400
+            type SerializedStateSize = digest::consts::U0;
```

Comment:
To properly implement this trait we need to implement optimized serialization for eager hashes (otherwise we would need to implement `ArraySize` for `U537` and `U569`). I plan to do it in later PRs.